### PR TITLE
`AssignAndPushReferenceValue` & `AppendAndPushReferenceValue` peephole opts

### PR DESCRIPTION
**Pull request by ike709**
When `AssignNoPush` is followed by `PushReferenceValue` and they both have the same reference, they're reduced to a single `Assign` with a new `AssignAndPushReferenceValue` optimization. This saves about ~17,000 redundant `ReadReference()` and `GetReferenceValue()` calls in TG bytecode.

I applied the same optimization to `AppendNoPush -> PushReferenceValue` when their refs are the same and shaved off 154 `PushReferenceValue`s in TG. Not nearly as impactful as the first opt, but peephole opts don't really cost us anything so I left it in.

### CODE REVIEW

**Comment by ZeWaka** [comment]
title and desc incorrect

**Comment by ike709** [comment]
waka lacking reading comprehension moment

File: DMCompiler/Optimizer/PeepholeOptimizations.cs [hunk 1]
```diff
@@ -775,4 +775,45 @@ public void Apply(List<IAnnotatedBytecode> input, int index) {
...
+// PushReferenceValue [ref]
+// -> Assign [ref]
+// These opcodes can be reduced to a single Assign as long as the [ref]s are the same
+internal sealed class AssignAndPushReferenceValue : IPeepholeOptimization {
+    public ReadOnlySpan<DreamProcOpcode> GetOpcodes() {
+        return [
+            DreamProcOpcode.AppendNoPush,
+            DreamProcOpcode.PushReferenceValue
+        ];
+    }
```
Comment (ike709) [inline DMCompiler/Optimizer/PeepholeOptimizations.cs:789]:
```suggestion
// AssignNoPush [ref]
// PushReferenceValue [ref]
// -> Assign [ref]
// These opcodes can be reduced to a single Assign as long as the [ref]s are the same
internal sealed class AssignAndPushReferenceValue : IPeepholeOptimization {
    public ReadOnlySpan<DreamProcOpcode> GetOpcodes() {
        return [
            DreamProcOpcode.AssignNoPush,
            DreamProcOpcode.PushReferenceValue
        ];
    }
```

**Comment by ZeWaka** [comment]
When are we getting `AssignAndAppendAndPushReferenceValueNTimesThenReturnNFloatListsThenPop` tho

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**